[Wang Zihan] iP

larrywang0701 opened the pull request:
# Duke
## Duke is a bot that can help you manage your tasks.
> Never forget your assessment deadlines again!
### The main features are:
 - Add tasks with type `todo`, `deadline`, `event` and `duration`
 - Mark tasks as completed or not
 - Find tasks by partial names
 - List all tasks
 - And more...
### How to use?
  1. Download the **latest released JAR file** [here](https://github.com/larrywang0701/ip/releases)
  2. **Run the JAR file** with command `java -jar <filename>`
  3. **Type your command** to the command input box and hit Enter or press "Send" button.
  4. Enjoy it! :smile:
### To-do
 - [X] Improve GUI
 - [ ] Release the final version
### For developers:
If you need to add a new command, you can add it here in class main.Main
```java
     /**
     * Default constructor of the program's main class
     */
    public Main() {
        this.parser = new Parser();
        this.taskList = Storage.createTaskListFromFile();
        this.parser.registerCommand("intro", new CommandIntroHandler());
        this.parser.registerCommand("list", new CommandListHandler());
        this.parser.registerCommand("add", new CommandAddTaskHandler());
        this.parser.registerCommand("bye", new CommandExitHandler());
        CommandMarkUnmarkHandler temp = new CommandMarkUnmarkHandler();
        this.parser.registerCommand("mark", temp);
        this.parser.registerCommand("unmark", temp);
        this.parser.registerCommand("todo", new CommandTodoHandler());
        this.parser.registerCommand("deadline", new CommandDeadlineHandler());
        this.parser.registerCommand("event", new CommandEventHandler());
        this.parser.registerCommand("delete", new CommandDeleteHandler());
        this.parser.registerCommand("find", new CommandFindHandler());
        this.parser.registerCommand("duration", new CommandDurationHandler());
        // Register your new command here...
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

larrywang0701 committed:
Level-0

larrywang0701 committed:
Level-2

larrywang0701 committed:
Level-2

larrywang0701 committed:
Level-3

larrywang0701 committed:
Level-4

larrywang0701 committed:
A-TextUiTesting

larrywang0701 committed:
Level-5

larrywang0701 committed:
Level-6

larrywang0701 committed:
Level-7

larrywang0701 committed:
Merge pull request #1 from larrywang0701/branch-Level-7

Level-7 Merge

larrywang0701 committed:
Level-8

larrywang0701 committed:
Merge pull request #2 from larrywang0701/branch-Level-8

Level-8 Merge

larrywang0701 committed:
A-MoreOOP

larrywang0701 committed:
Merge pull request #3 from larrywang0701/branch-A-MoreOOP

A-MoreOOP Merge

larrywang0701 committed:
Merge pull request #4 from larrywang0701/add-gradle-support

Merge gradle support branch

larrywang0701 committed:
A-JUnit

larrywang0701 committed:
A-Jar

larrywang0701 committed:
A-JavaDoc

larrywang0701 committed:
Finish task A-CodingStandard

larrywang0701 committed:
Add find command (for Level-9)

larrywang0701 committed:
Merge branch 'master' into branch-A-CodingStandard

larrywang0701 committed:
Merge pull request #5 from larrywang0701/branch-A-CodingStandard

Merge branch-A-CodingStandard

larrywang0701 committed:
Merge pull request #6 from larrywang0701/branch-Level-9

Add find command (for Level-9)

PetrichorPrecipice commented on src/main/java/main/Main.java:
```diff
@@ -0,0 +1,125 @@
+package main;
+
+import command.CommandAddTaskHandler;
+import command.CommandExitHandler;
+import command.CommandIntroHandler;
+import command.CommandListHandler;
+import command.CommandMarkUnmarkHandler;
+import command.CommandTodoHandler;
+import command.CommandDeadlineHandler;
+import command.CommandEventHandler;
+import command.CommandDeleteHandler;
+import command.CommandFindHandler;
+import command.Parser;
```
I like how you didn't import command.* and instead imported all the classes separately even though there were quite a few.

PetrichorPrecipice commented on src/main/java/command/CommandExitHandler.java:
```diff
@@ -0,0 +1,11 @@
+package command;
+import main.Main;
```
Should this import statement directly follow the package statement?

PetrichorPrecipice commented on src/main/java/task/TaskList.java:
```diff
@@ -0,0 +1,101 @@
+     * @return The task instance found by the given name, or "null" if no task in the list has the given name.
+     */
+    public Task findTaskByName(String name) {
+        int count = this.tasks.size();
```
```suggestion
        int count = tasks.size();
```
is "this" required here?

PetrichorPrecipice commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,55 @@
+     */
+    public Task(String name) {
+        this.name = name;
+        this.isDone = false;
```
```suggestion
        isDone = false;
```
Should "this" be here?

PetrichorPrecipice commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,55 @@
+     * @return task name
+     */
+    public String getName(){
+        return this.name;
```
```suggestion
        return name;
```
Should "this" be here?

PetrichorPrecipice commented on src/main/java/command/CommandMarkUnmarkHandler.java:
```diff
@@ -0,0 +1,24 @@
+package command;
+import main.Main;
```
```suggestion
package command;

import main.Main;
```
Should the import statement directly follow the package statement?

PetrichorPrecipice commented on src/main/java/main/Ui.java:
```diff
@@ -0,0 +1,63 @@
+     * The default constructor for the Ui class
+     */
+    public Ui() {
+        this.scanner = new Scanner(System.in);
```
```suggestion
        scanner = new Scanner(System.in);
```
Is "this" required here? I noticed the same issue in other parts of the code.

PetrichorPrecipice reviewed (commented):
Looks good to me! Code is succinct and well-written with just a few minor coding standard violations here and there. After fixing those up everything should be good.

larrywang0701 committed:
Replace text UI with simple GUI

WinSheng1 commented on README.md:
```diff
@@ -13,7 +13,7 @@ Prerequisites: JDK 11, update Intellij to the most recent version.
 1. Configure the project to use **JDK 11** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
    In the same dialog, set the **Project language level** field to the `SDK default` option.
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+3. After that, locate the `src/main/java/main/Main.java` file, right-click it, and choose `Run main.Main.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
```
I like how you changed the README, I think many of us (including me) didn't bother changing the filepath which is important for users who wants to use the application.

WinSheng1 commented on src/main/java/command/CommandAddTaskHandler.java:
```diff
@@ -0,0 +1,11 @@
+package command;
+import main.Main;
```
```suggestion
package command;

import main.Main;
```
Should this have a line separation in between package and import? Notice this in a few other areas.

WinSheng1 commented on src/main/java/command/CommandFindHandler.java:
```diff
@@ -0,0 +1,15 @@
+
+public class CommandFindHandler implements ICommandHandler {
+    @Override
+    public void execute(String command, String[] parameters) throws CommandException {
```
There should be a line separation here as well between the class and the method.

WinSheng1 commented on src/main/java/main/Main.java:
```diff
@@ -0,0 +1,125 @@
+import command.Parser;
+import task.TaskList;
+import util.Storage;
+
```
Should there be 3 empty lines between the import statements and the class?

WinSheng1 commented on src/main/java/main/Ui.java:
```diff
@@ -0,0 +1,63 @@
+     * @param outputLowerLine Set to true to output a split line under the message
+     */
+    public void say(String content, boolean outputUpperLine, boolean outputLowerLine) {
+        if(outputUpperLine){
```
```suggestion
        if (outputUpperLine) {
```
There should be spacings between the brackets. Noticed this in a few other areas as well.

WinSheng1 commented on src/main/java/task/EventTask.java:
```diff
@@ -0,0 +1,40 @@
+        this.from = DateTimeUtil.parseDateTimeString(fromStr);
+        this.to = DateTimeUtil.parseDateTimeString(toStr);
+        if(this.from.compareTo(this.to) > 0){
+            throw new CommandException("Start time " + DateTimeUtil.dateTimeToString(this.from) + " of this event is behind the end time " + DateTimeUtil.dateTimeToString(this.to) + ".");
```
Might want to check if this exceeds 120 characters.

WinSheng1 reviewed (commented):
Overall, your code follows most of Java's coding standards, with a few areas that can be improved - mainly separation of lines and spacings. Variable and method names are generally easy to understand and Javadocs are well described. Keep up the good work!

larrywang0701 committed:
Merge pull request #7 from larrywang0701/branch-Level-10

GUI (branch-Level-10) Merge

larrywang0701 committed:
Add assertions for checking null before executing commands

When executing many commands, I need to make sure that
`Main.getInstance()`, `Main.getInstance().getTaskList()` and
`Ui.getInstance()` are not null, otherwise the program will run into
bugs. So I added some assertions to first check whether these required
objects are null or not before executing command, so when these objects
are null when executing commands, the command won't continue executing
into a buggy result.

larrywang0701 committed:
Merge pull request #8 from larrywang0701/branch-A-Assertions

Add assertions for checking null before executing commands

larrywang0701 committed:
Improve code quality

larrywang0701 committed:
Add more JavaDoc

Add JavaDoc for more public methods and constructors

larrywang0701 committed:
Merge pull request #10 from larrywang0701/branch-A-CodeQuality

Merge branch-A-CodeQuality

larrywang0701 committed:
Add support for fixed duration tasks

Add a new command `duration` for managing fixed duration tasks with a
given duration in hours.

larrywang0701 committed:
Merge pull request #11 from larrywang0701/branch-B-FixedDurationTasks

Add support for fixed duration tasks

larrywang0701 committed:
Improve GUI

Add circular portrait image
Adjust text position and alignment
Error message highlighting

larrywang0701 committed:
Merge pull request #12 from larrywang0701/branch-A-BetterGui

Improve GUI

larrywang0701 committed:
Add user guide and UI screenshot

Add user guide in docs/README.md and UI screenshot in docs/Ui.png

larrywang0701 committed:
Add bot name at the top of the user guide

Add the bot's name "Duke" at the top of the user guide in docs/README.md

larrywang0701 committed:
Add error handling for exceptions when reading file

Add an error message box to tell the user there is an error and ask the user what to do when the program can not correctly load the task list from file.
